test/e2e_node/device_plugin_failures_pod_status_test.go

```diff
@@ -0,0 +1,256 @@
+		gomega.Expect(err).To(gomega.Succeed())
+
+		// wait for the pod to be deleted
+		gomega.Eventually(func() error {
+			_, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			return err
+		}, f.Timeouts.PodDelete, f.Timeouts.Poll).Should(gomega.MatchError((gomega.ContainSubstring("not found"))))
+	})
+
+	ginkgo.It("will report a Device Status for the failed pod in the pod status", func(ctx context.Context) {
+		// randomizing so tests can run in parallel
+		resourceName := fmt.Sprintf("test.device/%s", f.UniqueName)
+		devices := []kubeletdevicepluginv1beta1.Device{{ID: "testdevice", Health: kubeletdevicepluginv1beta1.Healthy}}
+		plugin := testdeviceplugin.NewDevicePlugin(nil)
+
+		err := plugin.RegisterDevicePlugin(ctx, f.UniqueName, resourceName, devices)
+		defer plugin.Stop() // should stop even if registration failed
+		gomega.Expect(err).To(gomega.Succeed())
+
+		ginkgo.By("initial state: capacity and allocatable are set")
+		gomega.Eventually(getNodeResourceValues, nodeStatusUpdateTimeout, f.Timeouts.Poll).WithContext(ctx).WithArguments(resourceName).Should(gomega.Equal(ResourceValue{Allocatable: 1, Capacity: 1}))
+
+		// schedule a pod that requests the device
+		client := e2epod.NewPodClient(f)
+		pod := client.Create(ctx, createPodWrongImage(resourceName, 1))
+
+		// wait for the pod to be running
+		gomega.Expect(e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "Back-off pulling image", f.Timeouts.PodStartShort,
+			func(pod *v1.Pod) (bool, error) {
+				if pod.Status.Phase == v1.PodPending &&
+					len(pod.Status.ContainerStatuses) > 0 &&
+					pod.Status.ContainerStatuses[0].State.Waiting != nil &&
+					pod.Status.ContainerStatuses[0].State.Waiting.Reason == "ImagePullBackOff" {
+					return true, nil
+				}
+				return false, nil
+			})).To(gomega.Succeed())
+
+		expectedStatus := []v1.ResourceStatus{
+			{
+				Name: v1.ResourceName(resourceName),
+				Resources: []v1.ResourceHealth{
+					{
+						ResourceID: "testdevice",
+						Health:     v1.ResourceHealthStatusHealthy,
+					},
+				},
+			},
+		}
+
+		gomega.Eventually(func() []v1.ResourceStatus {
+			pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			return pod.Status.ContainerStatuses[0].AllocatedResourcesStatus
+		}, devicePluginUpdateTimeout, f.Timeouts.Poll).Should(gomega.Equal(expectedStatus))
+
+		// now make the device unhealthy
+		devices[0].Health = kubeletdevicepluginv1beta1.Unhealthy
+		plugin.UpdateDevices(devices)
+
+		expectedStatus[0].Resources[0] = v1.ResourceHealth{
+			ResourceID: "testdevice",
+			Health:     "Unhealthy",
+		}
+
+		gomega.Eventually(func() []v1.ResourceStatus {
+			pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			return pod.Status.ContainerStatuses[0].AllocatedResourcesStatus
+		}, devicePluginUpdateTimeout, f.Timeouts.Poll).Should(gomega.Equal(expectedStatus))
```

Comment:
probalby material for beta, but should we update at all the resource status for pods which are in terminal state? I wonder what's the benefit here

Reply:
yes, that was the idea. Some pods may fail because of device unhealthy. Thus I want to update the status on failed pods

Reply:
ok, this can really wait to beta or later, but the reason I'm asking is: if a pod fails because of unhealhy devices, don't we have already (at admission time) all the info we need, and thus we can set the device status correctly right from the start?
In addition, if a pod goes failed for reasons not related to devices, and then the assigned devices incidentally go unhealthy, is it still correct/desirable to update the container statuses?
I guess the latter is really a corner case we can discuss pre-beta or even post-beta